pkg/apis/elasticsearch/v1/elasticsearch_types.go

```diff
@@ -72,8 +77,264 @@ type ElasticsearchSpec struct {
+
+	// Monitoring enables you to extract log and stack monitoring metrics of this Elasticsearch cluster.
+	// Metricbeat and Filebeat are deployed in the same pod as sidecar and each one send data to one or two different
+	// monitoring Elasticsearch clusters running in the same Kubernetes cluster.
+	// +kubebuilder:validation:Optional
+	Monitoring Monitoring `json:"monitoring,omitempty"`
+}
+
+// Monitoring holds Elasticsearch references to send logs and metrics of this Elasticsearch cluster.
+type Monitoring struct {
+	Metrics MetricsMonitoring `json:"metrics,omitempty"`
+	Logs    LogsMonitoring    `json:"logs,omitempty"`
+}
+
+type MetricsMonitoring struct {
+	// ElasticsearchRef is a reference to a monitoring Elasticsearch cluster running in the same Kubernetes cluster
+	// dedicated to receiving stack monitoring metrics.
+	ElasticsearchRef commonv1.ObjectSelector `json:"elasticsearchRef,omitempty"`
+}
+
+type LogsMonitoring struct {
+	// ElasticsearchRef is a reference to a monitoring Elasticsearch cluster running in the same Kubernetes cluster
+	// dedicated to receiving Elasticsearch logs.
+	ElasticsearchRef commonv1.ObjectSelector `json:"elasticsearchRef,omitempty"`
+}
+
+// MetricsMonitoringAssociation helps to manage the (Elasticsearch+Metricbeat)/Elasticsearch association
+type MetricsMonitoringAssociation struct {
+	*Elasticsearch
+	ref types.NamespacedName
+}
+
+var _ commonv1.Association = &MetricsMonitoringAssociation{}
+
+func (mma *MetricsMonitoringAssociation) Associated() commonv1.Associated {
+	if mma == nil {
+		return nil
+	}
+	if mma.Elasticsearch == nil {
+		mma.Elasticsearch = &Elasticsearch{}
+	}
+	return mma.Elasticsearch
+}
+
+func (mma *MetricsMonitoringAssociation) AssociationConfAnnotationName() string {
+	// annotation key should be stable to allow Elasticsearch Controller only pick up the ones it expects,
+	// based on ElasticsearchRefs
+
+	nsNameHash := sha256.New224()
+	// concat with dot to avoid collisions, as namespace can't contain dots
+	_, _ = nsNameHash.Write([]byte(fmt.Sprintf("%s.%s", mma.ref.Namespace, mma.ref.Name)))
+	// base32 to encode and limit the length, as using Sprintf with "%x" encodes with base16 which happens to
+	// give too long output
+	// no padding to avoid illegal '=' character in the annotation name
+	hash := base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(nsNameHash.Sum(nil))
+
+	return commonv1.FormatNameWithID(
+		commonv1.MetricbeatConfigAnnotationNameBase+"%s",
+		hash,
+	)
+}
+
+func (mma *MetricsMonitoringAssociation) AssociationType() commonv1.AssociationType {
+	return commonv1.MetricbeatAssociationType
+}
+
+func (mma *MetricsMonitoringAssociation) AssociationRef() commonv1.ObjectSelector {
+	return commonv1.ObjectSelector{
+		Name:      mma.ref.Name,
+		Namespace: mma.ref.Namespace,
+	}
+}
+
+func (mma *MetricsMonitoringAssociation) AssociationConf() *commonv1.AssociationConf {
+	if mma.assocConfs == nil {
```

Comment:
nil pointer deref if metrics are not defined due to a pointer to Elasticsearch being embedded in the struct

Reply:
I think it can't happen because when it's call, an association has always been created, but it doesn't hurt to add a nil check.

Note that we have the same code for the Agent.
https://github.com/elastic/cloud-on-k8s/blob/c0b259b8a29d64371a339539846bd42e2d58ddc6/pkg/apis/agent/v1alpha1/agent_types.go#L257-L258

Reply:
Yes but for agent the association is only created when needed. Just try to run it an example with log shipping but without metrics shipping. 

The problem is that you return an empty struct in `GetMonitoringMetricsAssociation` and `GetMonitoringLogsAssocation` but then access the embedded Elasticsearch pointer unconditionally here. 

```diff
--- a/pkg/apis/elasticsearch/v1/elasticsearch_types.go
+++ b/pkg/apis/elasticsearch/v1/elasticsearch_types.go
@@ -288,7 +288,7 @@ func (es *Elasticsearch) GetMonitoringMetricsAssociation() commonv1.Association
                        ref:           ref.WithDefaultNamespace(es.Namespace).NamespacedName(),
                }
        }
-       return &MetricsMonitoringAssociation{}
+       return &MetricsMonitoringAssociation{Elasticsearch: es}
 }
 
 func (es *Elasticsearch) GetMonitoringLogsAssociation() commonv1.Association {
@@ -299,7 +299,7 @@ func (es *Elasticsearch) GetMonitoringLogsAssociation() commonv1.Association {
                        ref:           ref.WithDefaultNamespace(es.Namespace).NamespacedName(),
                }
        }
-       return &LogsMonitoringAssociation{}
+       return &LogsMonitoringAssociation{Elasticsearch: es}
 }
```